### ISSUE TITLE
Uploading encoded images into linearly-encoded GPUTextures

### ISSUE THREAD

**Issue by kainino0x** [gpuweb/gpuweb#1715]
Image sources (like JPEG) are typically non-linearly encoded, e.g. with sRGB.
Using `copyExternalImageToTexture`, these non-linear values can be uploaded into:

- e.g. `rgba8unorm`, if the application wants to sample the original values
- e.g. `rgba8unorm-srgb`, if the application wants to sample linear values (e.g. for linear rendering pipelines).

However in order to store actual linear values in the texture (in particular with formats like `rgba16float`), currently applications would need to perform the conversion themselves (either on CPU + writeTexture, or in a compute/fragment shader).

A destination color-space option on `copyExternalImageToTexture` (soon to be added) would provide a place for users to control the upload: instead of `srgb`, they could specify `srgb-linear`, which is the same color space but linearly encoded.

**Comment by kainino0x** [comment]
It would be achieved using whatever "srgb-linear" thing CSS comes up with:
https://github.com/w3c/csswg-drafts/issues/6087

**Comment by svgeesus** [comment]
This is [in CSS Color 4 now](https://drafts.csswg.org/css-color-4/#predefined-sRGB-linear)

```css
/* these are the same color */

 color(srgb 0.691 0.139 0.259)
 color(srgb-linear 0.435 0.017 0.055)
```

**Comment by donmccurdy** [comment]
Noting that CSS includes Display P3 color space...

```
enum PredefinedColorSpace { "srgb", "display-p3" };
```

... can the texture upload path support this? For example, I think a wide-gamut linear rendering pipeline might need to do something like:

1. Download Display P3 texture
2. Upload to GPU as Linear Display P3 or perhaps ACEScg (_not_ Linear-sRGB)
3. Perform rendering calculations in ACEScg
4. Render output to Display P3 or sRGB, depending on device capabilities

**Comment by kainino0x** [comment]
Yes, it will be able to (but adding it hasn't been a priority yet since it's going to take a fair amount of implementation work and needs to be well tested). As an example, if you have a Display P3 encoded image:

- Create an ImageBitmap `ib` from it. `ib` is tagged as display-p3.
- copyExternalImageToTexture with the destination GPUImageCopyTextureTagged.colorSpace = 'display-p3-linear'. Image gets converted from display-p3 to display-p3-linear
- Operate in display-p3-linear space.
- Output to a canvas with GPUCanvasConfiguration.colorSpace = 'display-p3-linear'.

Alternatively if the application works in floating point (e.g. rgba16float instead of fixed point like rgba8unorm) then it can just operate in srgb-linear instead of display-p3-linear:

- copyExternalImageToTexture with the destination GPUImageCopyTextureTagged.colorSpace = 'srgb-linear'. Image gets converted from display-p3 to (extended) srgb-linear. May result in numbers outside 0-1 so the destination must be rgba16float.
- Operate in srgb-linear space.
- Output to a canvas tagged as GPUCanvasConfiguration.colorSpace = 'srgb-linear'.

**Comment by kainino0x** [comment]
'srgb-linear' and 'display-p3-linear' could be in GPUPredefinedColorSpace even if they're not in PredefinedColorSpace. I'm thinking we will continue to have our own enum because (1) it lets us add the linear spaces (which we need) and (2) it means that we don't suddenly have to support new spaces every time they get added in the HTML spec.
https://gpuweb.github.io/gpuweb/#enumdef-gpupredefinedcolorspace

**Comment by donmccurdy** [comment]
Thanks @kainino0x! It's helpful to know that's on the roadmap.

**Comment by kainino0x** [comment]
[1] A conversation on Dawn's implementation work leads me to propose that copyExternalImageToTexture into `-srgb` formats should only be allowed with destination `colorSpace`s which are srgb-encoded (so, "srgb" and "display-p3" are OK while "srgb-linear" and "prophoto-rgb" are not).

I'm pretty sure the cases this disallows are not generally useful. Disallowing them simplifies implementations by allowing copyExternalImageToTexture to write using a render pass and simply skipping the final srgb-re-encode in the color conversion code, rather than having to use a format-reinterpreting view or intermediate texture.

**Comment by donmccurdy** [comment]
I'm not quite sure I understand the restriction you're describing — is this a fair paraphrase? 

> When uploading textures into `*-srgb` formats like `rgba8unorm-srgb` the selected destination color space must use an sRGB transfer function (✅ sRGB, Display P3; 🚫 Linear-sRGB, AdobeRGB).

If so, I think that's a reasonable requirement. The workflows you describe in https://github.com/gpuweb/gpuweb/issues/1715#issuecomment-1003190377 would upload into `rgba8unorm` (not `rgba8unorm-srgb`) because they're using `display-p3-linear` or `srgb-linear` as the destination color space?

**Comment by kainino0x** [comment]
> is this a fair paraphrase?

Yup!

> The workflows you describe in [#1715 (comment)](https://github.com/gpuweb/gpuweb/issues/1715#issuecomment-1003190377) would upload into `rgba8unorm` (not `rgba8unorm-srgb`) because they're using `display-p3-linear` or `srgb-linear` as the destination color space?

You're correct the workflows I described wouldn't upload into rgba8unorm-srgb because they wouldn't want the srgb format semantics (gamma-encoding when stored and gamma-decoding when loaded). They might use either rgba8unorm or rgba16float depending on the precision they need.

**Comment by kainino0x** [comment]
[2] I had another realization that writing into `-srgb` formats doesn't make any sense when the destination has `premultipliedAlpha: true`. Premultiplied alpha doesn't work with `-srgb` formats because they specifically assume unpremultiplied data. So I think we can restrict that as well. This makes it _actually_ possible to cancel out of the following three steps in a fragment-shader-based implementation.

- conversion from destination-gamut-linear to destination-gamut-nonlinear
- premultiplication (which validation ensures does not occur)
- conversion from destination-gamut-nonlinear to destination-gamut-linear to cancel out the gamma-encoding done by the hardware when writing an `-srgb` format

**Comment by kainino0x** [comment]
> writing into `-srgb` formats doesn't make any sense when the destination has `premultipliedAlpha: true`

Thinking about it further, I guess this claim is too strong. I think it works just fine in practice even though the intermediate results are strange. I think it would still be OK to disallow, since I think allowing it has implementation cost and is likely a minor footgun.

Here's what I was thinking: it goes against how the -srgb formats are designed. The RGB channels are gamma-encoded on write and gamma-decoded on read, so you can store gamma-encoded data in memory but work with linear (non-gamma-encoded) numbers. When the hardware performs those implicit gamma encode/decode steps, it processes R, G, and B independently without considering A. So, for example, if you perform premultiplication before storing into an srgb texture, you represent a 50%-opaque full-white (1, 1, 1) as premultiplied RGBA (0.5, 0.5, 0.5, 0.5), then write that to an sRGB texture as (0.21, 0.21, 0.21, 0.5).

If instead you perform premultiplication AFTER conversion to srgb, you convert linear (1, 1, 1, 0.5) to srgb (1, 1, 1, 0.5), then premultiply to get (0.5, 0.5, 0.5, 0.5). This is what I'd expect because full-white (1, 1, 1) shouldn't be affected by sRGB, and it maintains the invariant that r=g=b=a is white, and that (r,g,b) &le; a, rather than (r,g,b) &le; gamma(a).
